src/pages/signup/signup.test.js

```diff
@@ -0,0 +1,149 @@
+import React from 'react';
+import {
+  fireEvent,
+  render,
+  screen,
+  waitFor,
+} from '@testing-library/react';
+import SignUp from 'pages/signup';
+import userData from 'users.json';
+import bcrypt from 'bcryptjs';
+
+const handleInput = (selector, value) => {
+  fireEvent.change(selector, {
+    target: {
+      value,
+    },
+  });
+};
+
+const fireSubmit = () => {
+  fireEvent.click(screen.getByRole('button', { name: 'Sign up' }));
+};
+
+const testCases = (type, dataCases) => {
+  const title = `Render ${type} $field error message`;
+  test.each(dataCases)(title, async ({ field, data, error }) => {
+    render(<SignUp />);
+
+    const inpField = screen.getByLabelText(field);
+    const inpData = data;
+
+    handleInput(inpField, inpData);
+    fireSubmit();
+
+    await waitFor(() => {
+      expect(screen.getByText(error)).toBeInTheDocument();
+    });
+  });
+};
+
```

Comment:
I think we should not do this, it may shorter your code but it is very hard to maintain, and upgrade in the future for customizing. And it also reduce the readability a lot, people look at the test cases will confuse what is this meaning

Reply:
I don't quite get what 'this' we're currently talking here, the preview image is just brackets at the end of the code

Reply:
<img width="776" alt="Screen Shot 2022-07-09 at 22 56 51" src="https://user-images.githubusercontent.com/107382512/178113235-f7e2a691-4819-4db2-859b-8562edea0d76.png">
Stupid GitHub :(